### PR TITLE
feat(resolver): Add CLI option to resolve minimal version dependencies

### LINKED ISSUE #4100: Command to update Cargo.lock to minimal versions
From today's meeting.

Problem:

You write a library A, which depends on B, so you put `B = 1.0` in A's Cargo.toml. The you run Cargo build, and Cargo greedily pulls B 1.1 into the lockfile. Then you accidentally start depending on features introduced in 1.1, but you don't change `Cargo.toml`. Your test locally pass, and CI passes as well, and you publish a crate whose Cargo.toml is a lie. 

Solution:

Add `cargo update --minimal`, which generates lockfile picking the minimum possible version of all crates (it's not possible, of course, because there's no total order on dependency graphs, but some heuristics might work well in practice). Then in CI environment you generate the minimal lockfile to make sure you don't accidentally depend on newer than Cargo.toml features.

### PR DESCRIPTION
**Pull request by klausi**
Fixes #4100 

Test cases are still missing. We need to come up with a plan what cases we want to cover.

Thanks a lot to @Eh2406 for very helpful instructions to kick this off.

### CODE REVIEW

**Comment by Eh2406** [comment]
Thanks for working on this! Let me know when this is ready for review, or if there is anything else I can do to help?

One complication is that @alexcrichton has recently https://github.com/rust-lang/cargo/pull/5180 add code that uses the sort to preserve the lock file. I think we should always have `in_previous` first; then only the `.summary.version()` should be subject to this switch. @alexcrichton what do you think?

Also where is @rust-highfive to set:
r? @alexcrichton

**Comment by alexcrichton** [comment]
> I think we should always have in_previous first; then only the .summary.version() should be subject to this switch. @alexcrichton what do you think?

Agreed!

**Comment by klausi** [comment]
Added a test case, this should now be ready for review.

I tried to follow your suggestion to only apply the switch when `in_previous` has been checked:

```rust
ret.sort_unstable_by(|a, b| {
    let a_in_previous = self.try_to_use.contains(a.summary.package_id());
    let b_in_previous = self.try_to_use.contains(b.summary.package_id());
    let previous_cmp = a_in_previous.cmp(&b_in_previous);
    match previous_cmp {
        Ordering::Equal => {
            let cmp = a.summary.version().cmp(&b.summary.version());
            match self.minimal_versions {
                true => cmp.reverse(),
                false => cmp,
            }
        },
        _ => previous_cmp,
    }
});
```

But that does not work and then my test case fails. Maybe I need to specify a more elaborate test case that would then drive around this `in_previous` check? Or maybe we always need to reverse() as it is in the PR changes right now?

**Comment by Eh2406** [comment]
Thanks for working on this. That looks like it should work. I wonder why the test fails with that code. I can experiment with it when I get home.

**Comment by Eh2406** [comment]
Got it the old code was `a.cmp(&b).reverse()`! So that means the new code needs a `.reverse()` at the end asswell. (or any number of equivalent ways of writing it, if it makes things clearer. For example the [`.then_with(`](https://doc.rust-lang.org/stable/std/cmp/enum.Ordering.html#method.then_with) method may be easier to read, and `match` on a bool is somtimes cleaner as an `if`.

**Comment by alexcrichton** [comment]
@bors: r+

This looks great, thanks @klausi!

**Comment by alexcrichton** [comment]
@bors: r-

oh? How so? (I think I missed that...)

**Comment by Eh2406** [comment]
https://github.com/rust-lang/cargo/pull/5200#issuecomment-374771936 suggested how to handle it correctly, but it was failing the test. So it was **Not** committed. https://github.com/rust-lang/cargo/pull/5200#issuecomment-374794220 figured out how to get it working, but that has not yet been incorporated.

After that is fixed there are some nits, working on a revue now, then we will be good.

File: tests/testsuite/resolve.rs [hunk 1]
```diff
@@ -320,6 +359,48 @@ fn test_resolving_maximum_version_with_transitive_deps() {
     assert_that(&res, is_not(contains(names(&[("util", "1.1.1")]))));
 }
 
+#[test]
+fn test_resolving_minimum_version_with_transitive_deps() {
```
Comment (Eh2406) [inline tests/testsuite/resolve.rs]: I really like this test, it concisely demonstrates both properties of  `minimal-versions`, that it chooses the smallest version while meeting all the requirements. Mabey a comment at the beginning explaining that for the next reader. @alexcrichton has been trying to teach me to leave more comments.

Also @alexcrichton, do you think this needs a full integration test that calls `cargo -Z minimal-versions` or is this sufficient?

File: src/cargo/core/resolver/mod.rs [hunk 2]
```diff
@@ -797,7 +807,10 @@ impl<'a> RegistryQueryer<'a> {
             let b_in_previous = self.try_to_use.contains(b.summary.package_id());
             let a = (a_in_previous, a.summary.version());
             let b = (b_in_previous, b.summary.version());
-            a.cmp(&b).reverse()
+            match self.minimal_versions {
+                true => a.cmp(&b),
+                false => a.cmp(&b).reverse(),
+            }
```
Comment (Eh2406) [inline src/cargo/core/resolver/mod.rs]: This needs your code from https://github.com/rust-lang/cargo/pull/5200#issuecomment-374771936. Also [match on a bool](https://rust-lang-nursery.github.io/rust-clippy/master/index.html#match_bool) is somtimes better as an `if`, and the [`.then_with(`](https://doc.rust-lang.org/stable/std/cmp/enum.Ordering.html#method.then_with) method may be helpfull. I leave it to your judgment.

**Comment by klausi** [comment]
Thanks a lot @Eh2406 , again :-)

Very good feedback, I think I covered your points now.

File: tests/testsuite/resolve.rs (hunk 1 above)
Comment (alexcrichton) [reply tests/testsuite/resolve.rs]: Oh sorry I missed this. Yeah @klausi mind adding a small integration test to ensure that the flag is plumbed to the right location?

**Comment by klausi** [comment]
And now with integration test. tests/testsuite/generate_lockfile.rs might not be the best place to put it, any better suggestions?

File: tests/testsuite/resolve.rs (hunk 1 above)
Comment (klausi) [reply tests/testsuite/resolve.rs]: done!

**Comment by klausi** [comment]
Moved the integration test right after the resolver test, I guess it makes sense to just have them together there.

**Comment by Eh2406** [comment]
This looks good to me, r+! Thank you @klausi.

**Pull request merged**